src/lib/x509/x509_ext.h

```diff
@@ -560,6 +562,252 @@ class BOTAN_PUBLIC_API(3, 5) TNAuthList final : public Certificate_Extension {
+                  if(val[static_cast<size_t>(V) - 1 - i] != 0) {
+                     break;
+                  }
+               }
+               return IPAddressBlocks::IPAddress<V>(val);
+            }
+
+            friend std::strong_ordering operator<=>(const IPAddress<V> lhs, const IPAddress<V>& rhs) {
+               for(size_t i = 0; i < static_cast<size_t>(V); i++) {
+                  if(lhs.value()[i] < rhs.value()[i]) {
+                     return std::strong_ordering::less;
+                  } else if(lhs.value()[i] > rhs.value()[i]) {
+                     return std::strong_ordering::greater;
+                  }
+               }
+               return std::strong_ordering::equal;
+            }
+
+            friend bool operator==(const IPAddress<V>& lhs, const IPAddress<V>& rhs) {
+               for(size_t i = 0; i < static_cast<size_t>(V); i++) {
+                  if(lhs.value()[i] != rhs.value()[i]) {
+                     return false;
+                  }
+               }
+               return true;
+            }
+
+         private:
+            std::array<uint8_t, static_cast<size_t>(V)> m_value;
+      };
+
+      template <Version V>
+      class BOTAN_PUBLIC_API(3, 7) IPAddressOrRange final : public ASN1_Object {
+         public:
+            void encode_into(DER_Encoder&) const override;
+            void decode_from(class BER_Decoder& from) override;
+
+            IPAddressOrRange() = default;
+
+            IPAddressOrRange(const IPAddress<V>& addr) : m_min(addr), m_max(addr) {}
+
+            IPAddressOrRange(const IPAddress<V>& min, const IPAddress<V>& max) : m_min(min), m_max(max) {}
+
+            IPAddress<V> min() const { return m_min; }
+
+            IPAddress<V> max() const { return m_max; }
+
+         private:
+            IPAddress<V> m_min;
+            IPAddress<V> m_max;
+      };
+
+      template <Version V>
+      class BOTAN_PUBLIC_API(3, 7) IPAddressChoice final : public ASN1_Object {
+         public:
+            void encode_into(DER_Encoder&) const override;
+            void decode_from(BER_Decoder& from) override;
+
+            const std::optional<std::vector<IPAddressOrRange<V>>>& ranges() const { return m_ip_addr_ranges; }
+
+            IPAddressChoice() = default;
+
+            IPAddressChoice(const std::vector<IPAddressOrRange<V>>& ranges) : m_ip_addr_ranges(ranges) {
+               sort_and_merge();
+            }
+
+         private:
+            std::optional<std::vector<IPAddressOrRange<V>>> m_ip_addr_ranges;
+            void sort_and_merge();
+      };
+
+      class BOTAN_PUBLIC_API(3, 7) IPAddressFamily final : public ASN1_Object {
+         public:
+            void encode_into(DER_Encoder&) const override;
+            void decode_from(BER_Decoder& from) override;
+
+            IPAddressFamily() = default;
+
+            IPAddressFamily(const std::variant<IPAddressChoice<Version::IPv4>, IPAddressChoice<Version::IPv6>>& choice,
```

Comment:
Add a typedef for the `std::variant` within `IPAddressFamily`, and use throughout